crates/ika-core/src/dwallet_mpc/session_request.rs

```diff
@@ -0,0 +1,975 @@
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+
+    DKGFirst {
+        curve: DWalletMPCNetworkKeyScheme,
+        dwallet_id: ObjectID,
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+
+    DKGSecond {
+        curve: DWalletMPCNetworkKeyScheme,
+        encrypted_centralized_secret_share_and_proof: Vec<u8>,
+        encryption_key: Vec<u8>,
+        dwallet_id: ObjectID,
+        encrypted_secret_share_id: ObjectID,
+        dwallet_network_encryption_key_id: ObjectID,
+        first_round_output: SerializedWrappedMPCPublicOutput,
+        centralized_public_key_share_and_proof: SerializedWrappedMPCPublicOutput,
+    },
+
+    Presign {
+        curve: DWalletMPCNetworkKeyScheme,
+        signature_algorithm: SignatureAlgorithm,
+        dwallet_id: Option<ObjectID>,
+        presign_id: ObjectID,
+        dwallet_public_output: Option<SerializedWrappedMPCPublicOutput>,
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+
+    Sign {
+        curve: DWalletMPCNetworkKeyScheme,
+        hash_scheme: Hash,
+        signature_algorithm: SignatureAlgorithm,
+        dwallet_id: ObjectID,
+        sign_id: ObjectID,
+        is_future_sign: bool,
+        dwallet_network_encryption_key_id: ObjectID,
+        dwallet_decentralized_public_output: SerializedWrappedMPCPublicOutput,
+        message: Vec<u8>,
+        presign: SerializedWrappedMPCPublicOutput,
+        message_centralized_signature: SerializedWrappedMPCPublicOutput,
+    },
+
+    NetworkEncryptionKeyDkg {
+        key_scheme: DWalletMPCNetworkKeyScheme,
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+
+    NetworkEncryptionKeyReconfiguration {
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+
+    EncryptedShareVerification {
+        curve: DWalletMPCNetworkKeyScheme,
+        encrypted_centralized_secret_share_and_proof: Vec<u8>,
+        decentralized_public_output: SerializedWrappedMPCPublicOutput,
+        encryption_key: Vec<u8>,
+        dwallet_id: ObjectID,
+        encrypted_user_secret_key_share_id: ObjectID,
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+
+    PartialSignatureVerification {
+        curve: DWalletMPCNetworkKeyScheme,
+        message: Vec<u8>,
+        hash_type: Hash,
+        signature_algorithm: SignatureAlgorithm,
+        dwallet_decentralized_output: SerializedWrappedMPCPublicOutput,
+        presign: SerializedWrappedMPCPublicOutput,
+        partially_signed_message: SerializedWrappedMPCPublicOutput,
+        dwallet_id: ObjectID,
+        partial_centralized_signed_message_id: ObjectID,
+        dwallet_network_encryption_key_id: ObjectID,
+    },
+}
+
+pub(crate) enum AdvanceSpecificData {
+    ImportedKeyVerification {
+        curve: DWalletMPCNetworkKeyScheme,
+        public_input: <DWalletImportedKeyVerificationParty as mpc::Party>::PublicInput,
+        encrypted_centralized_secret_share_and_proof: Vec<u8>,
+        encryption_key: Vec<u8>,
```

Comment:
This code duplication could be avoided as follows; each such variant would have its own `struct ImportedKeyVerificationData`, which holds the inner-data in the `ProtocolSpecificData` variant, which would simply wrap it: `ImportedKeyVerification(ImportedKeyVerificationData)`; in `AdvanceSpecificData`, we can have that struct plus the advance request: `ImportedKeyVerification{ data: ImportedKeyVerificationData, advance_request: AdvanceRequest<()> }`